### PR TITLE
[WIP] midi interface revamp

### LINKED ISSUE #62: Refresh MIDI Device list without restart
1) Opened rack with MIDI Device - No devices listed.
2) Launched CV Toolkit which provides a virtual MIDI Divice
3) MIDI Device in Rack device list is still empty.
4) Restart Rack
5) CoreMIDI: from CV Toolkit 1 (input) (. . .) now shows. 

Would like to refresh without restart.  Removing/re-adding the MIDI Device did not refresh list either.

### PR DESCRIPTION
**Pull request by bontric**
This implements the midi interface revamp as described in #81 

- [x] Compare rtMidi/portmidi

**Conclusion**: Switching to rtMidi fixes #62 and allows using multiple interfaces on a single port without any additional code. (As far as I can tell, portmidi would require an additional wrapper to handle multiple interfaces for a portmidi stream.)

- [x]  Update Makefile to switch from portmidi to rtMidi
**Note:** Works on linux, but this should be tested on other systems to assure it works (I'm not fluent with Makefiles, so I guessed. This might work or not, we'll see ) 

- [x] Switch MidiInterface implementation to rtMidi

# List from #81
- [ ] MIDI-to-CV
	- [ ] 1V/oct, gate, velocity, pitchwheel, mod, volume, aftertouch outputs
	- [ ] MIDI clock output, start/stop gate, song reset outputs
	- [ ] LED for each output
	- [x] Be able to use the two interfaces on the same port, useful for MIDI channels
- [ ] Quad MIDI-to-CV
	- [ ] polyphonic voice asignment modes (rotate, reset, reasign)
	- [ ] 4x 1V/oct, 4x gate, 4x velocity outputs
	- [ ] pitchwheel, mod, volume outputs
	- [ ] LED for each gate
- [ ]  MIDI CC-to-CV
	- [ ] grid of 16 CC voltage outputs
	- [ ] select source CC channel for each port by clicking the dropdown on each one
	- [ ] LEDs for each output
- [ ] CV-to-MIDI
	- [ ] 1V/oct, gate, velocity, pitchwheel, mod inputs
	- [ ] MIDI clock, start/stop gate, song reset inputs
	- [ ] LED for each input

### CODE REVIEW

**Comment by AndrewBelt** [comment]
Awesome! I'll test this on all three OS's in a sec.

Note your tabs/spaces

**Comment by AndrewBelt** [comment]
Hmm, having trouble compiling for Mac 10.7. I can compile it for 10.9, but I can't drop support for 10.7 and 10.8, so I'll have to figure out a way around this.

**Comment by bontric** [comment]
Cool, that was quick! 

This could be related:
https://github.com/superquadratic/rtmidi-python/issues/8

How do you run Astyle? I'll make sure to run it prior to pull requests in the future ;)

**Comment by AndrewBelt** [comment]
That issue isn't related. It looks like CFLAGS (containing the `-mmacosx-version-min=10.7` flag) isn't getting passed to some of the commands in the RtMidi Makefile.

You don't *need* to run Astyle, it's just how I fixed mixed tabs/spaces.

**Comment by AndrewBelt** [comment]
Good news, found the problem. The RtMidi build uses CPPFLAGS instead of CXXFLAGS, so after defining that, the stable RtMidi 3.0.0 version builds. Now I'm having trouble linking it against the Rack binary...

**Comment by AndrewBelt** [comment]
Figured it out, I was building RtMidi with libstdc++ but Rack uses libc++ on Mac, so I had to switch to libc++ for RtMidi.

Be sure to check out the latest source and rebuild the RtMidi dep.

**Comment by AndrewBelt** [comment]
Cool, and Windows works, so all is now good!

**Comment by bontric** [comment]
Great! 
Midi-To-CV is almost done (except for the clock.  I think I'll create a separate module for that since it might require a more sophisticated approach to avoid timing issues)

I'll add the MIDI CC-to-CV and make another pull request once done

**Comment by AndrewBelt** [comment]
Oh yes, not sure if that was clear, but there should be as many MIDI modules as we need, not just one. The "Core" puffin seems a bit empty at the moment. 

It might help to abstract a bit so reduce redundancy but whatever.

**Comment by bontric** [comment]
I've abstracted the channel and input/output device selection so far.  There might be more but I can add that later. I'll create another pull request soon, so you can give me some feedback on the structure

File: src/core/MidiInterface.cpp [hunk 1]
```diff
@@ -191,12 +179,13 @@ void MidiInterface::releaseNote(int note) {
 	}
 }
 
-void MidiInterface::processMidi(long msg) {
-	int channel = msg & 0xf;
-	int status = (msg >> 4) & 0xf;
-	int data1 = (msg >> 8) & 0xff;
-	int data2 = (msg >> 16) & 0xff;
-	// printf("channel %d status %d data1 %d data2 %d\n", channel, status, data1, data2);
+void MidiInterface::processMidi(std::vector<unsigned char> msg) {
```
Comment (IohannRabeson) [inline src/core/MidiInterface.cpp]: it is a detail, or maybe I miss something, but using a const reference for the parameter msg can avoid a copy.

Comment (AndrewBelt) [reply src/core/MidiInterface.cpp]: Who cares, this is preoptimization

Comment (IohannRabeson) [reply src/core/MidiInterface.cpp]: well, a detail yes, a preoptimization, not sure, just a language detail. But as you say, who cares.

Comment (AndrewBelt) [reply src/core/MidiInterface.cpp]: Yeah, I need to do this with all my std::strings, but no time for that yet. And it looks uglier  :)

**Pull request merged**